src/v/raft/compaction_coordinator.cc

```diff
@@ -0,0 +1,296 @@
+    // just in case of out-of-order messages
+    stat_mcco = std::max(stat_mcco, remote_mcco);
+    recalculate_mtro();
+    return true;
+}
+
+model::offset compaction_coordinator::get_max_tombstone_remove_offset() const {
+    return _mtro;
+}
+
+model::offset
+compaction_coordinator::get_local_max_cleanly_compacted_offset() const {
+    return _local_mcco;
+}
+
+compaction_coordinator::clock_t::duration
+compaction_coordinator::base_interval() const {
+    auto retention = _log->config().tombstone_retention_ms();
+    // Typically retention is 24h, set update interval to 1h
+    // (up to 1.5h due to jitter).
+    // Using half-max for disabled to avoid overflow in jitter.
+    clock_t::duration base_interval = retention
+                                        ? clock_t::duration(*retention / 24)
+                                        : clock_t::duration::max() / 2;
+    vlog(_logger.debug, "calculated base_interval as {}", base_interval);
+    return base_interval;
+}
+
+void compaction_coordinator::on_ntp_config_change() {
+    auto base = base_interval();
+    _local_mcco_query_timer_jitter = jitter_t{base};
+    _rpc_sender.update_retry_interval(base);
+    // rearm
+    if (_local_mcco_query_timer.armed()) {
+        _local_mcco_query_timer.cancel();
+        arm_timer();
+    }
+}
+
+void compaction_coordinator::update_local_mcco() {
+    on_local_mcco_update(_log->cleanly_compacted_prefix_offset());
+    arm_timer();
+}
+
+void compaction_coordinator::on_local_mcco_update(model::offset new_mcco) {
+    vlog(
+      _logger.debug,
+      "updating local MCCO from {} to {}",
+      _local_mcco,
+      new_mcco);
+    if (new_mcco == _local_mcco) {
+        if (_need_force_update) {
+            _need_force_update = false;
+        } else {
+            return;
+        }
+    }
+    if (new_mcco < _local_mcco) {
+        vlog(
+          _logger.debug,
+          "attempted to move local MCCO backwards from {} to {}",
+          _local_mcco,
+          new_mcco);
+        return;
+    }
+    _local_mcco = new_mcco;
+
+    if (is_leader()) {
+        recalculate_mtro();
+    } else {
+        send_local_mcco_to_leader();
+    }
+}
+
+void compaction_coordinator::send_local_mcco_to_leader() {
+    if (!_leader_id) {
+        vlog(
+          _logger.trace,
+          "cannot send MCCO:{} to leader, no leader",
+          _local_mcco);
+        return;
+    }
+    vlog(
+      _logger.trace,
+      "sending local MCCO: {} to leader: {}",
+      _local_mcco,
+      _leader_id);
+    ssx::background = _leader_sender.submit(
+      _rpc_sender.make_local_mcco_sender(*_leader_id, _local_mcco));
+}
+
+void compaction_coordinator::send_mtro_to_followers() {
+    for (const auto& [node_id, fstat] : _fstats) {
+        if (node_id == _self) {
+            continue;
+        }
+        ssx::background = fstat.mtro_sender->submit(
+          _rpc_sender.make_mtro_sender(node_id, _mtro));
+    }
+}
```

Comment:
Would it simpler if leader requests local MCCOs from the followers ? This way we would have communication in a single direction. Current implementation is different than rest of Raft protocol as this is the only one where the follower initiates the communication. Can leader do this periodically ? I think this way the design would be simplified